File: 0001-0500/0199. Binary Tree Right Side View.go
package main
// Medium
// Time:O(n), Space:O(w)
// https://leetcode.cn/problems/binary-tree-right-side-view/

func rightSideView(root *TreeNode) []int {
    res := []int{}
    if root == nil {
        return res
    }

    que := []*TreeNode{root}
    for len(que) > 0 {
        levelSize := len(que)

        // go through cur level
        for i := 0; i < levelSize; i++ {
            node := que[0]
            que = que[1:] // pop out the que

            if i == levelSize - 1 {
                res = append(res, node.Val)
            }

            if node.Left != nil {
                que = append(que, node.Left)
            }
            if node.Right != nil {
                que = append(que, node.Right)
            }
        }
    }
    return res
}

File: 0001-0500/0199. Binary Tree Right Side View.py
# Time:O(n), Space:O(w)

from collections import deque
from typing import Optional, List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if not root:
            return []

        res = []
        que = deque([root])

        while que:
            size = len(que)

            for i in range(size):
                node = que.popleft()

                if i == size - 1:
                    res.append(node.val)

                if node.left:
                    que.append(node.left)
                if node.right:
                    que.append(node.right)

        return res
